zaza/openstack/charm_tests/heat/tests.py

```diff
@@ -0,0 +1,355 @@
+                                        image.id, msg="glance image")
+
+    def _keypair_delete(self):
+        """Delete that keypair."""
+        logging.info('Deleting keypair...')
+        openstack_utils.delete_resource(self.nova_client.keypairs,
+                                        KEYPAIR_NAME, msg="nova keypair")
+
+    def test_100_domain_setup(self):
+        """Run required action for a working Heat unit."""
+        # Action is REQUIRED to run for a functioning heat deployment
+        logging.info('Running domain-setup action on heat unit...')
+        unit = zaza.model.get_units(self.application_name)[0]
+        assert unit.workload_status == "active"
+        zaza.model.run_action(unit.entity_id, "domain-setup")
+        zaza.model.block_until_unit_wl_status(unit.entity_id, "active")
+        unit = zaza.model.get_unit_from_name(unit.entity_id)
+        assert unit.workload_status == "active"
+
+    def test_400_heat_resource_types_list(self):
+        """Check default resource list behavior and confirm functionality."""
+        logging.info('Checking default heat resource list...')
+        try:
+            types = list(self.heat_client.resource_types.list())
+            if type(types) is list:
+                logging.info('Resource type list check is ok.')
+            else:
+                msg = 'Resource type list is not a list!'
+                logging.error('{}'.format(msg))
+                raise
+            if len(types) > 0:
+                logging.info('Resource type list is populated '
+                             '({}, ok).'.format(len(types)))
+            else:
+                msg = 'Resource type list length is zero!'
+                logging.error(msg)
+                raise
+        except Exception as e:
+            msg = 'Resource type list failed: {}'.format(e)
+            logging.error(msg)
+            raise
+
+    def test_402_heat_stack_list(self):
+        """Check default heat stack list behavior, confirm functionality."""
+        logging.info('Checking default heat stack list...')
+        try:
+            stacks = list(self.heat_client.stacks.list())
+            if type(stacks) is list:
+                logging.info("Stack list check is ok.")
+            else:
+                msg = 'Stack list returned something other than a list.'
+                logging.error(msg)
+                raise
+        except Exception as e:
+            msg = 'Heat stack list failed: {}'.format(e)
+            logging.error(msg)
+            raise
+
+    def test_410_heat_stack_create_delete(self):
+        """Create stack, confirm nova compute resource, delete stack."""
+        logging.info('Creating, deleting heat stack (compute)...')
+        self._image_create()
+        self._keypair_create()
+        self._stack_create()
+        self._stack_resource_compute()
+        self._stack_delete()
+        self._image_delete()
+        self._keypair_delete()
```

Comment:
Are any of these helper functions used outside of this test?  If not, then they should either be placed after the test, or just inlined into the test, please.

Reply:
@ajkavanagh I don't believe these functions are used anywhere else. However, I don't quite follow where they should be placed instead. You mean put them in a separate class/file like utils.py? Or just don't have separate functions and put all the code in the 410 function?

Reply:
The 'flow' in the file is awkward as the functions get defined before the are used; I'm questioning why they are functions at all; i.e. why aren't they just inline code in the test?  What value is gained from them being made into functions?